server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -108,174 +114,179 @@ protected boolean resolveIndex() {
-                return null;
+                primaryResponse = executionResult;
             }
+            context.finalize(primaryResponse);
         } else {
-            throw new AssertionError("unknown result type for " + request + ": " + operationResult.getResultType());
+            context.finalize(executionResult);
         }
+        assert context.isFinalized();
     }
 
-    /** Executes bulk item requests and handles request execution exceptions */
-    static Translog.Location executeBulkItemRequest(IndexMetaData metaData, IndexShard primary,
-                                                    BulkShardRequest request, Translog.Location location,
-                                                    int requestIndex, UpdateHelper updateHelper,
-                                                    LongSupplier nowInMillisSupplier,
-                                                    final MappingUpdatePerformer mappingUpdater) throws Exception {
-        final DocWriteRequest<?> itemRequest = request.items()[requestIndex].request();
-        final DocWriteRequest.OpType opType = itemRequest.opType();
-        final BulkItemResultHolder responseHolder;
-        switch (itemRequest.opType()) {
-            case CREATE:
-            case INDEX:
-                responseHolder = executeIndexRequest((IndexRequest) itemRequest,
-                        request.items()[requestIndex], primary, mappingUpdater);
-                break;
-            case UPDATE:
-                responseHolder = executeUpdateRequest((UpdateRequest) itemRequest, primary, metaData, request,
-                        requestIndex, updateHelper, nowInMillisSupplier, mappingUpdater);
-                break;
-            case DELETE:
-                responseHolder = executeDeleteRequest((DeleteRequest) itemRequest, request.items()[requestIndex], primary, mappingUpdater);
-                break;
-            default: throw new IllegalStateException("unexpected opType [" + itemRequest.opType() + "] found");
-        }
-
-        final BulkItemRequest replicaRequest = responseHolder.replicaRequest;
-
-        // update the bulk item request because update request execution can mutate the bulk item request
-        request.items()[requestIndex] = replicaRequest;
-
-        // Retrieve the primary response, and update the replica request with the primary's response
-        BulkItemResponse primaryResponse = createPrimaryResponse(responseHolder, opType, request);
-        if (primaryResponse != null) {
-            replicaRequest.setPrimaryResponse(primaryResponse);
+    private static UpdateHelper.Result translateRequestIfUpdate(PrimaryExecutionContext context, UpdateHelper updateHelper,
+                                                                LongSupplier nowInMillisSupplier) {
+        final DocWriteRequest.OpType opType = context.getCurrent().opType();
+        if (opType == DocWriteRequest.OpType.UPDATE) {
```

Comment:
This would mean the api of `executeDeleteRequestOnPrimary` and `executeIndexRequestOnPrimary` will be slightly more complicated, butI think it's worth it. I'll give it a shot.

Reply:
I tried to refactor translateRequestIfUpdate to as suggested, but it ended up being complicated as the method returns the update translation result and also converts it to an index/delete operation which is then put on the context. To do what was suggested, I would need to take the conversion logic and put it in `executeBulkItemRequest` but then that method will get too complicated. Bottom line, I couldn't find a way to make it work I was happy with.

Reply:
I've given this a try here: https://github.com/elastic/elasticsearch/commit/bbc5b1f25febaea8389773afa00542342970c9cc

I've pushed the main lifecycle into executeBulkItemRequest, and although this method is a little longer now, I find it easier to follow the lifecycle. There is a section delineating each stage, i.e., if you made it to here, then you're in TRANSLATED state, if you made it a bit further down, then you are in EXECUTED, etc. Let me know what you think.